pkg/trace/mapper_trace_parser.go

```diff
@@ -0,0 +1,128 @@
+package trace
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"os"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/vhive-serverless/loader/pkg/common"
+)
+
+type MapperTraceParser struct {
+	DirectoryPath         string
+	duration              int
+	functionNameGenerator *rand.Rand
+}
+
+type DeploymentInfo struct {
+	YamlLocation      string
+	PredeploymentPath []string
+}
+
+type MapperOutput struct {
+	ProxyFunction string `json:"proxy-function"`
+}
+
+type functionToDeploymentInfo map[string]DeploymentInfo
+
+type functionToProxy map[string]MapperOutput
+
+func NewMapperParser(directoryPath string, totalDuration int) *MapperTraceParser {
+	return &MapperTraceParser{
+		DirectoryPath: directoryPath,
+
+		duration:              totalDuration,
+		functionNameGenerator: rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+func (p *MapperTraceParser) extractFunctions(mapperOutput functionToProxy, deploymentInfo functionToDeploymentInfo, dirPath string) []*common.Function {
+	var result []*common.Function
+
+	invocations := parseInvocationTrace(dirPath+"/invocations.csv", p.duration)
+	runtime := parseRuntimeTrace(dirPath + "/durations.csv")
+	memory := parseMemoryTrace(dirPath + "/memory.csv")
+
+	runtimeByHashFunction := createRuntimeMap(runtime)
+	memoryByHashFunction := createMemoryMap(memory)
+
+	for i := 0; i < len(*invocations); i++ {
+		invocationStats := (*invocations)[i]
+		hashFunction := invocationStats.HashFunction
+		hashApp := invocationStats.HashApp
+		hashOwner := invocationStats.HashOwner
+		proxyFunction := mapperOutput[hashFunction+hashOwner+hashApp].ProxyFunction
+		yamlPath := deploymentInfo[proxyFunction].YamlLocation
+		predeploymentPath := deploymentInfo[proxyFunction].PredeploymentPath
+		function := &common.Function{
+			Name: fmt.Sprintf("%s-%d-%d", proxyFunction, i, p.functionNameGenerator.Uint64()),
+
+			InvocationStats:   &invocationStats,
+			RuntimeStats:      runtimeByHashFunction[hashFunction],
+			MemoryStats:       memoryByHashFunction[hashFunction],
+			YAMLPath:          yamlPath,
+			PredeploymentPath: predeploymentPath,
+		}
+
+		result = append(result, function)
+	}
+
+	return result
+}
+
+func (p *MapperTraceParser) Parse() []*common.Function {
```

Comment:
Should all `log.Warn` in this function be `log.Fatal`? If you cannot get the configuration, the loader should terminate, right?

Reply:
Yes, you are correct. I fixed this.